DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,197 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DataStructures.Lists 
+{
+    public class CircularBuffer<T> : IEnumerable<T> , ICollection<T> where T : IComparable<T> {
+        private T[] _circularBuffer;
+        private int _end;
+        private int _start;
+        private int _length { get => _circularBuffer.Length - 1; }
+        private readonly T _default;
+        private static readonly int _defaultBufferLength = 10;
+
+        /// <summary>
+        /// Controls whether data should be overridden when it is continously inserted without reading
+        /// </summary>
+        public bool CanOverride { get; }
+        public bool IsEmpty { get => _end == _start; }
+        public bool IsFilledUp { get => ((_end + 1) % _circularBuffer.Length == _start) && !_circularBuffer[_start].Equals(_circularBuffer[_end]); }
+
+
+        /// <summary>
+        /// Initializes a circular buffer with initial length of 10
+        /// </summary>
+        public CircularBuffer(bool canOverride=true) : this(_defaultBufferLength, canOverride) 
+        {
+        }
+
+        /// <summary>
+        /// Initializes a circular buffer with given length
+        /// </summary>
+        /// <param name="length">The length of the buffer</param>
+        public CircularBuffer(int length, bool canOverride=true) 
+        {
+            if (length < 1) 
+            {
+                throw new ArgumentOutOfRangeException("length can not be zero or negative");
+            }
+            _circularBuffer = new T[length + 1];
+            _default = _circularBuffer[length];
+            _end = 0;
+            _start = 0;
+            CanOverride = canOverride;
+        }
+
+        /// <summary>
+        /// Writes value to the back of the buffer
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Add(T value) 
+        {
+            if (CanOverride) 
+            {
+                innerInsert(value);
+            }
+            else 
+            {
+                if (IsFilledUp) 
+                {
+                    throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+                }
+                innerInsert(value);
+            }
```

Comment:
This method can be simplified with a single if-else clause, which would only contain a single call to `innerInsert` method, i.e.:
```
If(CanOverride == false && IsFilledUp == true)
{
    // throw new exception
}

innerInsert(value)
```

Reply:
This refactoring is ingeneous. Wonder why I didn't think of it :)